### PR TITLE
Emit warning if both tabularcolumns and :widths: are given (refs: #4196)

### PR DESCRIPTION
**Pull request by tk0miya**
refs: #4196

### CODE REVIEW

File: CHANGES [hunk 1]
```diff
@@ -43,6 +43,7 @@ Features added
 * #4168: improve zh search with jieba
 * HTML themes can set up default sidebars through ``theme.conf``
 * #3160: html: Use ``<kdb>`` to represent ``:kbd:`` role
+* #4196: latex: Emit warning if both tabularcolumns and :widths: are given
```
Comment (jfbu) [inline CHANGES]: perhaps use ``:rst:dir:`tabularcolumns` `` and  ```` ``:widths:`` ```` markup

File: sphinx/writers/latex.py [hunk 2]
```diff
@@ -1332,6 +1332,9 @@ def visit_table(self, node):
         self.table = Table(node)
         if self.next_table_colspec:
             self.table.colspec = '{%s}\n' % self.next_table_colspec
+            if 'colwidths-given' in node.get('classes', []):
+                logger.warning('both tabularcolumns and :widths: option are given. '
+                               ':withds: is ignored.', location=node)
```
Comment (jfbu) [inline sphinx/writers/latex.py]: there is a typo width

**Review by jfbu — CHANGES_REQUESTED** [review]
I wonder if we really need this: it is to be expected that one can not both obey ``:widths:`` and ``tabularcolumns``. Perhaps rather we should add mention of this behaviour in markup/misc.rst ?

**Comment by tk0miya** [comment]
Updated. I feel it is bad if sphinx ignores markups silently. So I'd like to emit warnings to let users know markups are wrong and ignored.

**Review by jfbu — APPROVED** [review]
The issue here is that ``:widths:`` makes senses and is obeyed by HTML. It is also obeyed by LaTeX in absence of ``tabularcolumns``.  Adding ``tabularcolumns`` is described as optional way for LaTeX only to override colspec. In my opinion, current behaviour is compliant with docs. When preparing a document both for HTML and LaTeX output I may now get warnings on LaTeX builds which tells nothing really new to the Sphinx user having read documentation. But I don't object strongly merging :-)

**Comment by jfbu** [comment]
I think in such case there is need for an "informational" level of warnings, which would not be treated as an error even under sphinx-build  ``-W`` option.

**Comment by tk0miya** [comment]
Thank you for comment. I changed the warning to an info message.
I'll merge this after tests passed

File: sphinx/writers/latex.py [hunk 3]
```diff
@@ -1376,6 +1376,9 @@ def visit_table(self, node):
         self.table = Table(node)
         if self.next_table_colspec:
             self.table.colspec = '{%s}\n' % self.next_table_colspec
+            if 'colwidths-given' in node.get('classes', []):
+                logger.info('both tabularcolumns and :widths: option are given. '
+                            ':widths: is ignored.', location=node)
```
Comment (jfbu) [inline sphinx/writers/latex.py:1381]: are you planning to make this translatable string?

**Review by jfbu — APPROVED** [review]
LGTM

File: sphinx/writers/latex.py (hunk 3 above)
Comment (tk0miya) [reply sphinx/writers/latex.py:1381]: oh I'd forgotten that.

Anyway, Sphinx only translates strings on documents, so these messages are not translated.
`__(...)` (double underscore) helper is provided, but it is not implemented now.
https://github.com/sphinx-doc/sphinx/blob/122a404eb3cf0c50f5ada60fc236d8b460c32492/sphinx/locale/__init__.py#L251-L258

**Comment by tk0miya** [comment]
merged.

**Comment by tk0miya** [comment]
Thank you for reviewing :-)

**Pull request merged**